File: app/src/main/java/com/example/nonameapp/base/BaseActivity.java
package com.example.nonameapp.base;

import android.os.Bundle;
import android.view.LayoutInflater;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewbinding.ViewBinding;

public abstract class BaseActivity<VB extends ViewBinding> extends AppCompatActivity {
    private final BindingInflater<VB> bindingInflater;
    private VB binding;

    protected BaseActivity(BindingInflater<VB> bindingInflater) {
        this.bindingInflater = bindingInflater;
    }

    protected VB getBinding() {
        return binding;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = bindingInflater.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        initData();
        bindData();
        setOnClick();
    }

    protected abstract void initData();

    protected abstract void bindData();

    protected abstract void setOnClick();

    public interface BindingInflater<VB extends ViewBinding> {
        VB inflate(LayoutInflater inflater);
    }
}

File: app/src/main/java/com/example/nonameapp/base/BaseActivity.kt
package com.example.nonameapp.base

import android.os.Bundle
import android.view.LayoutInflater
import androidx.appcompat.app.AppCompatActivity
import androidx.viewbinding.ViewBinding
import com.example.nonameapp.ui.SignUpFragment

abstract class BaseActivity<VB : ViewBinding>(
    private val bindingInflater: (LayoutInflater) -> VB
) : AppCompatActivity() {

    private var _binding: VB? = null
    protected val binding: VB
        get() = _binding as VB

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = bindingInflater(layoutInflater)
        setContentView(binding.root)

        initData()
        bindData()
        setOnClick()
    }

    abstract fun initData()

    abstract fun bindData()

    abstract fun setOnClick()
}
